drivers/i2c/i2c_nrfx_twim_rtio.c

```diff
@@ -0,0 +1,274 @@
+						    sqe->buf_len, dt_spec->addr);
+	case RTIO_OP_TINY_TX:
+		return i2c_nrfx_twim_rtio_msg_start(dev, I2C_MSG_WRITE | sqe->iodev_flags,
+						    sqe->tiny_buf, sqe->tiny_buf_len,
+						    dt_spec->addr);
+	case RTIO_OP_TX:
+		/* If buffer is not accessible by DMA, copy it into the internal driver buffer */
+		if (!nrf_dma_accessible_check(&config->common.twim, sqe->buf)) {
+			/* Validate buffer will fit */
+			if (sqe->buf_len > config->common.msg_buf_size) {
+				LOG_ERR("Need to use the internal driver "
+					"buffer but its size is insufficient "
+					"(%u > %u). "
+					"Adjust the zephyr,concat-buf-size or "
+					"zephyr,flash-buf-max-size property "
+					"(the one with greater value) in the "
+					"\"%s\"' node.",
+					sqe->buf_len, config->common.msg_buf_size, dev->name);
+				return i2c_rtio_complete(ctx, -ENOSPC);
+			}
+			memcpy(config->common.msg_buf, sqe->buf, sqe->buf_len);
+			sqe->buf = config->common.msg_buf;
+		}
+		return i2c_nrfx_twim_rtio_msg_start(dev, I2C_MSG_WRITE | sqe->iodev_flags, sqe->buf,
+						    sqe->buf_len, dt_spec->addr);
+	case RTIO_OP_I2C_CONFIGURE:
+		(void)i2c_nrfx_twim_configure(dev, sqe->i2c_config);
+		return false;
+	case RTIO_OP_I2C_RECOVER:
+		(void)i2c_rtio_recover(ctx);
+		return false;
+	default:
+		LOG_ERR("Invalid op code %d for submission %p\n", sqe->op, (void *)sqe);
+		return i2c_rtio_complete(ctx, -EINVAL);
+	}
+}
+
+static void i2c_nrfx_twim_rtio_complete(const struct device *dev, int status)
+{
+	/** Finalize if there are no more pending xfers */
+	const struct i2c_nrfx_twim_rtio_config *config = dev->config;
+	struct i2c_rtio *ctx = config->ctx;
+
+	if (i2c_rtio_complete(ctx, status)) {
+		(void)i2c_nrfx_twim_rtio_start(dev);
+	} else {
+		/* Release bus on completion */
+		pm_device_runtime_put(dev);
+	}
+}
+
+static int i2c_nrfx_twim_rtio_transfer(const struct device *dev, struct i2c_msg *msgs,
+				       uint8_t num_msgs, uint16_t addr)
+{
+	const struct i2c_nrfx_twim_rtio_config *config = dev->config;
+	struct i2c_rtio *ctx = config->ctx;
+
+	return i2c_rtio_transfer(ctx, msgs, num_msgs, addr);
+}
+
+static void i2c_nrfx_twim_rtio_submit(const struct device *dev, struct rtio_iodev_sqe *iodev_seq)
+{
+	const struct i2c_nrfx_twim_rtio_config *config = dev->config;
+	struct i2c_rtio *ctx = config->ctx;
+
+	if (i2c_rtio_submit(ctx, iodev_seq)) {
+		if (pm_device_runtime_get(dev) < 0) {
+			(void)i2c_rtio_complete(ctx, -EINVAL);
+		} else {
+			(void)i2c_nrfx_twim_rtio_start(dev);
+		}
+	}
+}
+
+static void event_handler(nrfx_twim_evt_t const *p_event, void *p_context)
+{
+	const struct device *dev = p_context;
+	int status = p_event->type == NRFX_TWIM_EVT_DONE ? 0 : -EIO;
+
+	i2c_nrfx_twim_rtio_complete(dev, status);
+}
+
+static const struct i2c_driver_api i2c_nrfx_twim_driver_api = {
+	.configure = i2c_nrfx_twim_configure,
+	.transfer = i2c_nrfx_twim_rtio_transfer,
+	.recover_bus = i2c_nrfx_twim_recover_bus,
```

Comment:
This should point to a wrapper that calls i2c_rtio_recover which would enqueue the recovery command rather than directly doing the recovery, without this the ordering of operations may result in a race

Reply:
I'm guessing this also applies to the existing TWI driver, which I copied this from?